pkg/controller.v2/controller_service.go

```diff
@@ -40,115 +41,110 @@ func (tc *TFJobController) reconcileServices(
-		diffIndexes := getDiffServiceIndexes(activeServices, *spec.Replicas)
-		if diff+len(diffIndexes) != 0 {
-			// This should never happened.
-			return fmt.Errorf("services diff(%d) is not equal to length(%d) of diffIndexes", diff, len(diffIndexes))
-		}
-
-		expectationServicesKey := genExpectationServicesKey(tfjobKey, rt)
-		err := tc.expectations.ExpectCreations(expectationServicesKey, diff)
-		if err != nil {
-			return err
-		}
+	replicas := int(*spec.Replicas)
+	// Get all pods for the type rt.
+	services = filterServicesForTFReplicaType(services, rt)
 
-		for _, index := range diffIndexes {
-			log.Infof("need to create new service: %s-%s", rt, index)
-
-			// Create OwnerReference.
-			controllerRef := genOwnerReference(tfjob)
-
-			// Append tfReplicaTypeLabel and tfReplicaIndexLabel labels.
-			labels := genLabels(tfjobKey)
-			labels[tfReplicaTypeLabel] = rt
-			labels[tfReplicaIndexLabel] = index
-
-			service := &v1.Service{
-				Spec: v1.ServiceSpec{
-					Selector: labels,
-					Ports: []v1.ServicePort{
-						{
-							Name: genGeneralName(tfjobKey, rt, index),
-							Port: defaultServicePort,
-						},
-					},
-				},
-			}
+	serviceSlices := getServiceSlices(services, replicas, loggerForReplica(tfjob, rt))
 
-			service.Name = genGeneralName(tfjobKey, rt, index)
-			service.Labels = labels
-
-			err := tc.serviceControl.CreateServicesWithControllerRef(tfjob.Namespace, service, tfjob, controllerRef)
-			if err != nil && errors.IsTimeout(err) {
-				// Service is created but its initialization has timed out.
-				// If the initialization is successful eventually, the
-				// controller will observe the creation via the informer.
-				// If the initialization fails, or if the pod keeps
-				// uninitialized for a long time, the informer will not
-				// receive any update, and the controller will create a new
-				// pod when the expectation expires.
-				return nil
-			} else if err != nil {
+	for index, serviceSlice := range serviceSlices {
+		if len(serviceSlice) > 1 {
+			loggerForReplica(tfjob, rt).Warningf("We have to many services for %s %d", rt, index)
+			// TODO(gaocegege): Kill some services.
+		} else if len(serviceSlice) == 0 {
+			loggerForReplica(tfjob, rt).Infof("need to create new service: %s-%d", rt, index)
+			err := tc.createNewService(tfjob, rt, strconv.Itoa(index), spec)
+			if err != nil {
 				return err
 			}
 		}
-	} else if diff > 0 {
-		// TODO(CPH): Need to delete service.
-		log.Infof("need to delete service but it is not implemented yet")
 	}
 
 	return nil
 }
 
-// getDiffServiceIndexes checks and gets diff indexes from desired and current.
-func getDiffServiceIndexes(activeServices []*v1.Service, replicas int32) []string {
-	desiredIndexes := make(map[string]string)
-
-	for i := int32(0); i < replicas; i++ {
-		desiredIndexes[fmt.Sprintf("%d", i)] = noHit
-	}
-
-	for _, service := range activeServices {
+// getServiceSlices returns a slice, which element is the slice of service.
```

Comment:
Comment isn't all that clear. It looks like the return value is a list of list so it would be good to explain what each dimension corresponds to e.g.
result[i] - Is an array of pointers to services corresponding to Services for replica i